### PR TITLE
API rework

### PR DESCRIPTION
**Pull request by vhf**
As requested by @bergos

### CODE REVIEW

**Comment by herrstucki** [comment]
I'm curious about the motivation for this. If we follow the RDF Data Cube Spec, a DataSet is a "A collection of statistical data that corresponds to a defined structure", which is definitely how we are thinking about data both from a provider and consumer's side. Also, I've never heard anyone use the term Data Cube for anything concrete in practice.

IMO the property of having datasets with one _defined structure_ (per spec: DataStructureDefinition) each is really important. I think renaming them to data cubes is quite confusing.

**Comment by vhf** [comment]
@bergos and @ktk will explain their reasoning. :)

**Comment by ktk** [comment]
I promised some reasoning for that and I plan to merge some of this information into the readme later.

We all worked with RDF Data Cubes so for us `DataSet` makes sense. But I had concerns from the first version of the API, see the now closed issue #3. I did argue in the end that we can leave it as is but Thomas had the same concerns so we re-visited it.

RDF Data Cubes are just an implementation of the generic [OLAP](https://en.wikipedia.org/wiki/Online_analytical_processing) concept and more specifically, [OLAP cube](https://en.wikipedia.org/wiki/OLAP_cube). Statistik Stadt Zürich for example is using [SAS](https://www.sas.com/de_ch/software/olap.html#m=features), have a look at the terms they use on this "feature" page. You will not find the term _data set_ there but just the more generic term _data cube_.

What RDF Data Cube calls _data structure definition_ can be mapped to _dimension tables_ in [OLAP cubes](https://en.wikipedia.org/wiki/OLAP_cube#Terminology), the _observations_ in RDF Data Cube are called _fact tables_. 

I understand that for all of us that so far worked with RDF Data Cubes this can be confusing but I'm quite sure for those coming from classical OLAP software like SAS it makes more sense. Right now our API is still pretty much tied to SPARQL and RDF Data Cubes but I could imagine that we will build another abstraction on top of it that is even closer to what "OLAP" people would expect, especially those that never heard of RDF and SPARQL.

Last but not least, people around Google/schema.org are interested in proving Data Cube like vocabulary within schema.org and they are not supper happy with the terms & structures used in RDF Data Cubes. So there is a possibility that schema.org will introduce a replacement vocab at one point that is using different terms more aligned with classical OLAP cubes.

Hope this clarifies it a bit.

**Comment by herrstucki** [comment]
@ktk Thanks for the explanation and I see better where this is coming from. Maybe as a counterpoint, I'll say that in my professional experience I have never encountered "Data Cube" as a term in any project, apart from one instance where the data cube was simply what the client called their database. They were still talking about datasets when referring to concrete … well, datasets.

For people _not_ coming from the OLAP world, I think "data set" or "dataset" is the much more approachable and [colloquial term](https://books.google.com/ngrams/graph?content=datacube%2Cdata+cube%2Cdataset%2Cdata+set&case_insensitive=on&year_start=1900&year_end=2000&corpus=15&smoothing=3&share=&direct_url=t4%3B%2Cdatacube%3B%2Cc0%3B%2Cs0%3B%3BDatacube%3B%2Cc0%3B%3Bdatacube%3B%2Cc0%3B%3BDataCube%3B%2Cc0%3B%3BDATACUBE%3B%2Cc0%3B.t4%3B%2Cdata%20cube%3B%2Cc0%3B%2Cs0%3B%3Bdata%20cube%3B%2Cc0%3B%3BData%20Cube%3B%2Cc0%3B%3BData%20cube%3B%2Cc0%3B.t4%3B%2Cdataset%3B%2Cc0%3B%2Cs0%3B%3Bdataset%3B%2Cc0%3B%3BDataset%3B%2Cc0%3B%3BDATASET%3B%2Cc0%3B%3BDataSet%3B%2Cc0%3B.t4%3B%2Cdata%20set%3B%2Cc0%3B%2Cs0%3B%3Bdata%20set%3B%2Cc0%3B%3BData%20Set%3B%2Cc0%3B%3BDATA%20SET%3B%2Cc0%3B%3BData%20set%3B%2Cc0#t4%3B%2Cdatacube%3B%2Cc0%3B%2Cs0%3B%3BDatacube%3B%2Cc0%3B%3Bdatacube%3B%2Cc0%3B%3BDataCube%3B%2Cc0%3B%3BDATACUBE%3B%2Cc0%3B.t4%3B%2Cdata%20cube%3B%2Cc0%3B%2Cs0%3B%3Bdata%20cube%3B%2Cc0%3B%3BData%20Cube%3B%2Cc0%3B%3BData%20cube%3B%2Cc0%3B.t4%3B%2Cdataset%3B%2Cc0%3B%2Cs0%3B%3Bdataset%3B%2Cc0%3B%3BDataset%3B%2Cc0%3B%3BDATASET%3B%2Cc0%3B%3BDataSet%3B%2Cc0%3B.t4%3B%2Cdata%20set%3B%2Cc0%3B%2Cs0%3B%3Bdata%20set%3B%2Cc0%3B%3BData%20Set%3B%2Cc0%3B%3BDATA%20SET%3B%2Cc0%3B%3BData%20set%3B%2Cc0). People use it for data in Excel sheets, CSV files, on paper etc. it has actually nothing to do with how it's named in the RDF Data Cube spec – that just happens to describe it accurately IMO.

In conclusion I think it's wrong to cater terminology to OLAP, which is just one specific flavor of data storage/modelling/processing. I'd much prefer to call a duck a duck, or a dataset a dataset.

Anyhow, that's just my 2 cents 😄 

PS: On the other hand, it doesn't really matter to us currently – we can of course use the library regardless if datasets are called datasets, data cubes, treasure chests, or chocolate croissants 😉

**Comment by herrstucki** [comment]
Also another thing that just came to mind: it's just right now in practice confusing to call things differently than in the RDF Data Cube spec, since the library – at least for the time being – is specifically designed to work with RDF Data Cubes (hence the name :)). Maybe save the naming change to a future version which aligns with a new Schema.org spec, when it's actually released? Or you could still spin off a more generalistic non-RDF specific library once you need that …

**Comment by ktk** [comment]
@herrstucki fair points. The OLAP world calls 2-dimensional views "slices" so one could argue that a dataset is a 2-dimensional data cube.

My proposal is that we leave everything as is now, get some more experience (we have enough projects for that in the pipeline ;) and re-visit this discussion sometime in 2020.

One possible outcome might be that we depreciate this package and refactor and rebrand it to whatever makes sense by then. Maybe the API will look different as well, it still feels like SPARQL right now and I'm not sure if this makes sense for non-RDF people.

**Comment by herrstucki** [comment]
@ktk Sounds good!
